extensions/wikidata/src/org/openrefine/wikidata/commands/ConnectionManager.java

```diff
@@ -0,0 +1,219 @@
+     */
+    public static final String PREFERENCE_STORE_KEY = "wikidata_credentials";
+
+    public static final int CONNECT_TIMEOUT = 5000;
+    public static final int READ_TIMEOUT = 10000;
+
+    /**
+     * For now, this class is hard-coded for Wikidata.
+     * <p>
+     * It will be generalized to work against other Wikibase instances in the future.
+     */
+    private static final String WIKIBASE_API_ENDPOINT = ApiConnection.URL_WIKIDATA_API;
+
+    /**
+     * The single {@link ApiConnection} instance managed by {@link ConnectionManager}.
+     * <p>
+     * Currently, only one connection is supported at the same time.
+     */
+    private ApiConnection connection;
+
+    private static final ConnectionManager instance = new ConnectionManager();
+
+    public static ConnectionManager getInstance() {
+        return instance;
+    }
+
+    private ConnectionManager() {
+        PreferenceStore prefStore = ProjectManager.singleton.getPreferenceStore();
+        // remove credentials stored in the preferences
+        prefStore.put(PREFERENCE_STORE_KEY, null);
+    }
+
+    /**
+     * Logs in to the Wikibase instance, using username/password.
+     * <p>
+     * If failed to login, the connection will be set to null.
+     *
+     * @param username the username to log in with
+     * @param password the password to log in with
+     * @return true if logged in successfully, false otherwise
+     */
+    public boolean login(String username, String password) {
+        connection = new BasicApiConnection(WIKIBASE_API_ENDPOINT);
+        setupConnection(connection);
+        try {
+            ((BasicApiConnection) connection).login(username, password);
+            return true;
+        } catch (LoginFailedException e) {
+            logger.error(e.getMessage());
+            connection = null;
+            return false;
+        }
+    }
+
+    /**
+     * Logs in to the Wikibase instance, using owner-only consumer.
+     * <p>
+     * If failed to login, the connection will be set to null.
+     *
+     * @param consumerToken  consumer token of an owner-only consumer
+     * @param consumerSecret consumer secret of an owner-only consumer
+     * @param accessToken    access token of an owner-only consumer
+     * @param accessSecret   access secret of an owner-only consumer
+     * @return true if logged in successfully, false otherwise
+     */
+    public boolean login(String consumerToken, String consumerSecret,
+                         String accessToken, String accessSecret) {
+        connection = new OAuthApiConnection(WIKIBASE_API_ENDPOINT,
+                consumerToken, consumerSecret,
+                accessToken, accessSecret);
+        setupConnection(connection);
+        try {
+            // check if the credentials are valid
+            connection.checkCredentials();
+            return true;
+        } catch (IOException | MediaWikiApiErrorException e) {
+            logger.error(e.getMessage());
```

Comment:
Is this message displayed to the user? It looks like it just gets put into the server log.

Reply:
Yes, only in server log. If the login fails, the frontend can know by checking the response and then asks the user to try again.